system-contracts/contracts/BatchAggregator.sol

```diff
@@ -0,0 +1,354 @@
+        /// Check State Diffs
+        /// encoding is as follows:
+        /// header (1 byte version, 3 bytes total len of compressed, 1 byte enumeration index size)
+        /// body (`compressedStateDiffSize` bytes, 4 bytes number of state diffs, `numberOfStateDiffs` * `STATE_DIFF_ENTRY_SIZE` bytes for the uncompressed state diffs)
+        /// encoded state diffs: [20bytes address][32bytes key][32bytes derived key][8bytes enum index][32bytes initial value][32bytes final value]
+        require(
+            uint256(uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]))) ==
+                STATE_DIFF_COMPRESSION_VERSION_NUMBER,
+            "state diff compression version mismatch"
+        );
+        uint256 stateDiffSliceStart = calldataPtr;
+        calldataPtr++;
+
+        uint24 compressedStateDiffSize = uint24(bytes3(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 3]));
+        calldataPtr += 3;
+
+        uint8 enumerationIndexSize = uint8(bytes1(_totalL2ToL1PubdataAndStateDiffs[calldataPtr]));
+        calldataPtr++;
+
+        bytes calldata compressedStateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            compressedStateDiffSize];
+        calldataPtr += compressedStateDiffSize;
+
+        bytes calldata totalL2ToL1Pubdata = _totalL2ToL1PubdataAndStateDiffs[:calldataPtr];
+
+        uint32 numberOfStateDiffs = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
+        calldataPtr += 4;
+
+        bytes calldata stateDiffs = _totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr +
+            (numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE)];
+
+        calldataPtr += numberOfStateDiffs * STATE_DIFF_ENTRY_SIZE;
+
+        repackStateDiffs(chainId, numberOfStateDiffs, enumerationIndexSize, stateDiffs, compressedStateDiffs);
+    }
+    function bytesLength(uint256 value) internal returns (uint8 length){
+        while(value>0){
+            length += 1;
+            value = value>>8;
+        }
+    }
+    
+    function compressValue(uint256 initialValue, uint256 finalValue) internal returns (bytes memory compressedDiff){
+        uint8 transform = bytesLength(finalValue);
+        uint8 add = bytesLength(finalValue-initialValue);
+        uint8 sub = bytesLength(initialValue-finalValue);
+        uint8 opt = (transform<add?transform:add);
+        opt = (opt<sub?opt:sub);
+        compressedDiff = new bytes(opt+1);
+        uint8 mask = (opt<<LENGTH_BITS_OFFSET);
+        uint256 value;
+        if (transform==opt){
+            mask |= 3;
+            value = finalValue;
+        }
+        else if (add==opt){
+            mask |= 1;
+            value = finalValue-initialValue;
+        }
+        else if (sub==opt){
+            mask |= 2;
+            value = initialValue-finalValue;
+        }
+        else{
+            require(false, "Optimal operation is not transform, add or sub");
```

Comment:
there are four types of compression:

```
    /// 0 -> Nothing (32 bytes)
    /// 1 -> Add
    /// 2 -> Subtract
    /// 3 -> Transform (< 32 bytes)
```

I am not sure this will work when the size of the change is 32 bytes

Reply:
It can't be 32 bytes. At most 2^32/2 = 2^31, because of add or subtract

Reply:
32 is bytes, not bits